### PR TITLE
fix(astro): prerendering issue when path contains underscore

### PR DESCRIPTION
**Pull request by V3RON**
https://github.com/withastro/astro/blob/769f11f23f3dd32da912d0688008f9233ae13dd4/packages/astro/src/core/util.ts#L122-L129

Notice the condition in line 124. It checks for the presence of `_` in any part of the URL. The problem is that file is an absolute path, so `_GITHUB` is picked up by this condition, causing the page to be considered 'not public'.

## Changes

Updated isPublicRoute to use relative paths, ensuring correct handling of underscore folders outside the repository. This change ensures the function correctly returns 'true' when an underscore folder is present outside the repository folder, as intended.

Closes https://github.com/withastro/astro/issues/11232

## Testing

Added automated test to cover edge cases involving underscores in paths.

## Docs

No update needed

### CODE REVIEW

**Comment by V3RON** [comment]
Quick note: packages/astro/test/fixtures/_underscore in folder name/node_modules/fake-astro-library/ is there on purpose to simulate case when page is exposed by a third-party library.

File: packages/astro/src/core/util.ts [hunk 1]
```diff
@@ -120,11 +120,23 @@ function isInjectedRoute(file: URL, settings: AstroSettings) {
...
 
 function isPublicRoute(file: URL, config: AstroConfig): boolean {
-	const pagesDir = resolvePages(config);
-	const parts = file.toString().replace(pagesDir.toString(), '').split('/').slice(1);
+	const rootDir = config.root.toString();
+	const pagesDir = resolvePages(config).toString();
+	const fileDir = file.toString();
+
+	// Normalize the file directory path by removing the pagesDir prefix if it exists,
+	// otherwise remove the rootDir prefix.
```
Comment (ematipico) [inline packages/astro/src/core/util.ts:128]: The comment should also explain **why** this is necessary. The code itself isn't too complicated to understand, so a comment that *explains* seems superfluous. However, it's worth adding an explain of the motive, because if you left a comment here, it means there's a special case to handle :)

File: .changeset/honest-ravens-double.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+"astro": patch
+---
+
+Fixes prerendering issue for libraries in node_modules when a folder with an underscore is in the path.
```
Comment (ematipico) [inline .changeset/honest-ravens-double.md:5]:
```suggestion
Fixes a prerendering issue for libraries in `node_modules` when a folder with an underscore is in the path.
```

**Review by ematipico — APPROVED** [review]

File: packages/astro/src/core/util.ts (hunk 1 above)
Comment (V3RON) [reply packages/astro/src/core/util.ts:128]: > It might happen that a page is located in `node_modules`, and the Astro project itself is in a path containing the `_` character. In such cases, we should use a relative path to bypass the segment containing `_`.

WDYT?

**Review by matthewp — APPROVED** [review]
great test, thanks!

**Comment by kirso** [comment]
I am facing the same issue, after changing to `hybrid`, I don't get any html pages in my .dist folder at all. Although I wonder if its the same reason. 

The way how I found out is by adding a server endpoint (with `prerender=false`) and my Pagefind now fails the build because it doesn't index any .html files.

I checked the solution here, but I don't have any folders with the `_` before it so I wonder if it is what is causing an issue.

The repo is here: https://github.com/kirso/KirillSo.com
